src/BullOak.Repositories.EventStore/EventstoreRepository.cs

```diff
@@ -9,28 +9,74 @@
 
     public class EventStoreRepository<TId, TState> : IStartSessions<TId, TState>
     {
-        private static readonly Task<bool> falseResult = Task.FromResult(false);
         private readonly IHoldAllConfiguration configs;
-        private readonly IEventStoreConnection connection;
+        private readonly Func<IEventStoreConnection> connectionFactory;
 
-        public EventStoreRepository(IHoldAllConfiguration configs, IEventStoreConnection connection)
+        public EventStoreRepository(IHoldAllConfiguration configs, Func<IEventStoreConnection> connectionFactory)
         {
-            this.configs = configs ?? throw new ArgumentNullException(nameof(connection));
-            this.connection = connection ?? throw new ArgumentNullException(nameof(connection));
+            this.configs = configs ?? throw new ArgumentNullException(nameof(configs));
+            this.connectionFactory = connectionFactory ?? throw new ArgumentNullException(nameof(connectionFactory));
         }
 
         public async Task<IManageSessionOf<TState>> BeginSessionFor(TId id, bool throwIfNotExists = false)
         {
-            if (throwIfNotExists && !(await Contains(id)))
-                throw new StreamNotFoundException(id.ToString());
+            IEventStoreConnection connection = null;
+            EventStoreSession<TState> session;
+            try
+            {
+                try
+                {
+                    connection = connectionFactory();
+                }
+                catch (Exception ex)
+                {
+                    throw new RepositoryUnavailableException(
+                        "Couldn't connect to the EvenStore repository. See InnerException for details", ex);
+                }
+
+                if (connection == null)
+                {
+                    throw new RepositoryUnavailableException(
+                        "Couldn't connect to the EvenStore repository. Connection object is null.");
+                }
+
+                if (throwIfNotExists && !await Contains(id, connection))
+                {
+                    throw new StreamNotFoundException(id.ToString());
+                }
 
-            var session = new EventStoreSession<TState>(configs, connection, id.ToString());
-            await session.Initialize();
+                session = new EventStoreSession<TState>(configs, connection, id.ToString());
+                await session.Initialize();
+
+            }
+            catch
+            {
+                CleanupConnection(connection);
+                throw;
+            }
 
             return session;
         }
 
-        public async Task<bool> Contains(TId selector)
+        private void CleanupConnection(IEventStoreConnection connection)
```

Comment:
This cleanup function could potentially be shared between EventStoreRepository and EventStoreSession

Reply:
Done